### PR TITLE
Add support for network device limits.priority option

### PR DESCRIPTION
**Pull request by mihalicyn**
lxd/device: add support for network device limits.priority option

This is a replacement feature for per-instance limits.network.priority option.
New approach does not require netprio cgroup to be suppored (it's from legacy
cgroup v1) and also it allows to set priority for virtual machine instances.

The recommended Linux kernel version (for full functionality) is 5.16.

### CODE REVIEW

File: lxd/instance/drivers/driver_lxc.go [hunk 1]
```diff
@@ -2564,6 +2564,27 @@ func (d *lxc) onStart(_ map[string]string) error {
 	if d.expandedConfig["limits.network.priority"] != "" {
 		go func(d *lxc) {
 			d.fromHook = false
+
+			timeout := 10 * time.Second
```
Comment (roosterfish) [inline lxd/instance/drivers/driver_lxc.go:2568]: Maybe it makes sense to use the same timeout pattern across the codebase. Tom once posted a good suggestion to safe resources here https://github.com/canonical/lxd/pull/11993#discussion_r1262190102. The final code is here https://github.com/canonical/lxd/blob/main/lxd/instance/drivers/driver_qemu.go#L7529

File: lxd/instance/drivers/driver_lxc.go [hunk 2]
```diff
@@ -7913,8 +7938,46 @@ func (d *lxc) removeDiskDevices() error {
 	return nil
 }
 
-// Network I/O limits.
-func (d *lxc) setNetworkPriority() error {
+func setNetworkPriorityWithNetfilter(d *lxc, networkPriority int) error {
+	// in setNetworkPriorityWithCgroup we set skb->priority on all
```
Comment (roosterfish) [inline lxd/instance/drivers/driver_lxc.go:7942]: Wouldn't parts of this comment be better placed on top of the func?

File: lxd/instance/drivers/driver_lxc.go [hunk 3]
```diff
@@ -7913,8 +7938,46 @@ func (d *lxc) removeDiskDevices() error {
...
+	for devName, devConfig := range d.ExpandedDevices() {
+		if devConfig["type"] != "nic" {
+			continue
+		}
+
+		hostDevName := d.ExpandedConfig()[fmt.Sprintf("volatile.%s.host_name", devName)]
+
+		var err error
+
+		err = d.state.Firewall.InstanceClearNetPrio(d.Project().Name, d.Name(), hostDevName)
```
Comment (roosterfish) [inline lxd/instance/drivers/driver_lxc.go:7961]:
```suggestion
		err := d.state.Firewall.InstanceClearNetPrio(d.Project().Name, d.Name(), hostDevName)
```

This can be simplified a bit.

File: lxd/instance/drivers/driver_lxc.go [hunk 4]
```diff
@@ -7913,8 +7938,46 @@ func (d *lxc) removeDiskDevices() error {
...
+		if err == nil && networkPriority != 0 {
+			err = d.state.Firewall.InstanceSetupNetPrio(d.Project().Name, d.Name(), hostDevName, networkPriority)
+		}
+
+		if err == nil {
```
Comment (roosterfish) [inline lxd/instance/drivers/driver_lxc.go:7966]: If there are multiple devices with type `nic` and setting the net prio fails for all of them, will it only display the error message of the last device? 
Would it be possible to return here right away if it fails for one of the devices?

**Review by roosterfish — CHANGES_REQUESTED** [review]
Looks good, just a few questions and remarks :)

File: lxd/instance/drivers/driver_lxc.go (hunk 1 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:2568]: Absolutely agree. Will do.

File: lxd/instance/drivers/driver_lxc.go (hunk 4 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:7966]: Yes, of course. But I've just followed existing logic from `setNetworkPriorityWithCgroup` (previous name `setNetworkPriority`). As far as I understand the idea here is that failing netprio setting is not very critical and we are comfortable with the case when network priority was successfully set on at least one device.

File: lxd/instance/drivers/driver_lxc.go (hunk 2 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:7942]: sure, will do.

**Comment by mihalicyn** [comment]
> Looks good, just a few questions and remarks :)

Thanks for your review, Julian! ;-)

File: lxd/instance/drivers/driver_lxc.go [hunk 5]
```diff
@@ -2564,6 +2564,27 @@ func (d *lxc) onStart(_ map[string]string) error {
 	if d.expandedConfig["limits.network.priority"] != "" {
 		go func(d *lxc) {
 			d.fromHook = false
+
+			timeout := 10 * time.Second
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+
+			for {
+				if d.IsRunning() {
```
Comment (masnax) [inline lxd/instance/drivers/driver_lxc.go:2573]:
```go
for !d.IsRunning() {
```

would save you four lines :)

File: lxd/instance/drivers/driver_lxc.go (hunk 4 above)
Comment (masnax) [reply lxd/instance/drivers/driver_lxc.go:7966]: Just a design opinion, but I usually don't like nesting logic in `if err == nil` conditions, instead I go with the more conventional `if err != nil` pattern, but log the error and `continue` instead of returning the error. I find this reads more consistently, and has the added benefit of logging the additional errors in case they're of interest.

```go
NetworkPriorityWithNetfilter(d *lxc, networkPriority int) error {
    errs := make([]error, 0, len(d.ExpandedDevices())
    for devName, devConfig := range d.ExpandedDevices() {
        if devConfig["type"] != "nic" {
          continue
        }

        hostDevName := d.ExpandedConfig()[fmt.Sprintf("volatile.%s.host_name", devName)]

	err := d.state.Firewall.InstanceClearNetPrio(d.Project().Name, d.Name(), hostDevName)
        if err != nil {
            errs = append(errs, err)
            logger.Warn("Failed to clear instance network priority", logger.Ctx{"name": d.Name(), "project": d.Project().Name(), "error": err}
            
            continue
        }
       
        if networkPriority == 0 {
            continue
        }
 
        err = d.state.Firewall.InstanceSetupNetPrio(d.Project().Name, d.Name(), hostDevName, networkPriority)
        if err != nil {
            errs = append(errs, err)
            logger.Warn("Failed to setup instance network priority", logger.Ctx{"name": d.Name(), "project": d.Project().Name(), "error": err}

            continue
        }
   }

   if len(errs) > 0 {
      return fmt.Errorf("Failed to set network priority for %d devices. Last error: %w", len(eers), errs[len(errs) - 1])
   }

   return nil
}

```

Comment (masnax) [reply lxd/instance/drivers/driver_lxc.go:7966]: Ah shoot my condition at the end is wrong, It will fail if there are any errors. I'll add back the `succes` string :)

Comment (masnax) [reply lxd/instance/drivers/driver_lxc.go:7966]: One difference in the my implementation is that the `if !succes` case will potentially return a `nil` if `d.ExpandedDevices()` is empty. In your case: 

```go
if !success {
		return fmt.Errorf("Failed to set network device priority: %s", lastError)
	}
```

You create a new error, but `lastError` can still be `nil` if `d.ExpandedDevices()` is empty. Do we want this to error out if there are no devices, or just silently fail to set anything and move on?

File: lxd/instance/drivers/driver_lxc.go [hunk 6]
```diff
@@ -7934,8 +7938,46 @@ func (d *lxc) removeDiskDevices() error {
...
+
+		if err == nil {
+			success = true
+		} else {
+			lastError = err
+		}
+	}
+
+	if !success {
+		return fmt.Errorf("Failed to set network device priority: %s", lastError)
```
Comment (masnax) [inline lxd/instance/drivers/driver_lxc.go:7974]: You should wrap this error with `%w`.

**Review by masnax — CHANGES_REQUESTED** [review]
lgtm!, just a question and one small nit :)

File: lxd/instance/drivers/driver_lxc.go (hunk 6 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go]: Fixed. Thanks!

File: lxd/instance/drivers/driver_lxc.go [hunk 7]
```diff
@@ -2564,6 +2564,19 @@ func (d *lxc) onStart(_ map[string]string) error {
 	if d.expandedConfig["limits.network.priority"] != "" {
 		go func(d *lxc) {
 			d.fromHook = false
+
+			timeout := time.Now().Add(10 * time.Second)
+
+			for !d.IsRunning() {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:2570]: Why is this needed?

File: lxd/devices.go [hunk 8]
```diff
@@ -689,7 +689,7 @@ func deviceEventListener(stateFunc func() *state.State) {
 
 			s := stateFunc()
 
-			if !s.OS.CGInfo.Supports(cgroup.NetPrio, nil) {
+			if e[1] != "add" {
```
Comment (tomponline) [inline lxd/devices.go:695]: Could you add a comment here explaining why we ignore events other than add? Thanks

File: lxd/instance/drivers/driver_lxc.go (hunk 7 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:2570]: container should be fully started at the point where we try to configure network priority,
because in `setNetworkPriorityWithCgroup` we are accessing the cgroup stuff and LXC API.

File: lxd/instance/drivers/driver_lxc.go [hunk 9]
```diff
@@ -7926,8 +7930,52 @@ func (d *lxc) removeDiskDevices() error {
...
+	// skb's that comes from the container cgroup on the enter to
+	// the host interfaces
+	// here we go another way, we set skb->priority on all skb's
+	// which exit the container network namespace once skb exits
+	// from the host peer of the container network device.
+
+	// Check that we at least succeeded to set an entry
+	success := false
+	var err error
+	for devName, devConfig := range d.ExpandedDevices() {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:7945]: As this is handled at a per-NIC device level and is set explicitly rather than in relation to other instances, is this something we can move into the per-NIC device drivers rather than handling in the lxc driver?

A bit like we do with the call to `err = d.state.Firewall.InstanceSetupBridgeFilter`?

File: lxd/instance/drivers/driver_lxc.go (hunk 7 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:2570]: mmm yeah, but was this not working before?

File: lxd/devices.go (hunk 8 above)
Comment (mihalicyn) [reply lxd/devices.go:695]: As you can see this change in a separate commit

File: lxd/instance/drivers/driver_lxc.go (hunk 7 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:2570]: yes. It was not working. I was surprised too.

Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:2570]: You can easily reproduce that by making container stop (with network limit set) and then start the container. Limit won't be set.

File: lxd/devices.go (hunk 8 above)
Comment (tomponline) [reply lxd/devices.go:695]: Yeah I agree, I thought a comment in the code would be useful for future readers.

File: lxd/instance/drivers/driver_lxc.go (hunk 7 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:2570]: Hrm ok, let me think about this for a bit as it seems like maybe we should trigger this event later rather than polling for it the instance being ready.

File: lxd/instance/drivers/driver_lxc.go (hunk 9 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:7945]: >As this is handled at a per-NIC device level and is set explicitly rather than in relation to other instances, is this something we can move into the per-NIC device drivers rather than handling in the lxc driver?

I thought about that too.

- this logic is not specific to any NIC driver, it's universal
- an old `setNetworkPriorityWithCgroup` is specific to the containers and not applicable to the virtual machines
- theoretically, we can make `setNetworkPriorityWithNetfilter` work for VMs too

this is a good question where I should place this but I decided to keep it in LXC driver because originally this feature was for containers and I decided to avoid more global changes as we can do it in the future if it's needed.

**Comment by tomponline** [comment]
@mihalicyn whilst we think about the netfilter netprio stuff, do you want to split the bug fixes for the cgroup functionality into a separate PR we can merge for LXD 5.17?

File: lxd/instance/drivers/driver_lxc.go (hunk 7 above)
Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:2570]: > Is this something we can add a test for in the test suite?

theoretically yes, to check that we need to dump firewall rules and parse them OR (in cgroup case) find cgroup and check netprio cgroup file content.

**Comment by mihalicyn** [comment]
> @mihalicyn whilst we think about the netfilter netprio stuff, do you want to split the bug fixes for the cgroup functionality into a separate PR we can merge for LXD 5.17?

sure as you say. I just thought that this thing is simple enough to be merged at once.

File: lxd/instance/drivers/driver_lxc.go (hunk 9 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:7945]: I wonder if there is even a need to keep the old cgroup approach now that we can do it via netfilter?

Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:7932]: > I wonder if there is even a need to keep the old cgroup approach now that we can do it via netfilter?

I think it's a good idea to get rid of it. As cgroup-v1 will die eventually. But it's a good question if we want to do this right now or make this transition smoother.

Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:7932]: the netfilter approach works with cgroupv1 and cgroupv2 systems right?

Comment (mihalicyn) [reply lxd/instance/drivers/driver_lxc.go:7932]: sure. netfilter does not involve cgroups at all

**Comment by tomponline** [comment]
@mihalicyn when you're ready for another review please can you click the refresh review button on my name top right so that I get a notification and it'll appear on my todo list on https://github.com/pulls/review-requested

Thanks

File: doc/reference/devices_nic.md [hunk 10]
```diff
@@ -442,6 +442,7 @@ Key                     | Type    | Default           | Description
 `limits.egress`         | string  | -                 | I/O limit in bit/s for outgoing traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.ingress`        | string  | -                 | I/O limit in bit/s for incoming traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.max`            | string  | -                 | I/O limit in bit/s for both incoming and outgoing traffic (same as setting both `limits.ingress` and `limits.egress`)
+`limits.priority`       | integer | -                 | `skb->priority` for outgoing traffic
```
Comment (tomponline) [inline doc/reference/devices_nic.md:445]: I think we will need to clarify what this config item does in more accessible style, as I doubt many of our users will know what "skb->priority" is. Instead can you expand this to explain what it does as an end result, and whether the priority increases with a higher or lower number? We should also include the range of allowed integers here.

File: lxd/device/device_utils_network.go [hunk 11]
```diff
@@ -560,6 +560,40 @@ func networkSetupHostVethLimits(m deviceConfig.Device) error {
 		}
 	}
 
+	networkPriority := 0
+	if m["limits.priority"] != "" {
+		networkPriority, err = strconv.Atoi(m["limits.priority"])
+		if err != nil {
+			return fmt.Errorf("Failed to parse limits.priority (%s): %s", m["limits.priority"], err)
```
Comment (tomponline) [inline lxd/device/device_utils_network.go:567]: We tend to use `%q` and drop the `()` for input strings that have an error, that way Go will wrap them in double quotes so differentiating from the error message and the actual input value is clear and consistent.

File: lxd/device/device_utils_network.go [hunk 12]
```diff
@@ -560,6 +560,40 @@ func networkSetupHostVethLimits(m deviceConfig.Device) error {
...
+
+	err = d.state.Firewall.InstanceClearNetPrio(d.inst.Project().Name, d.inst.Name(), veth)
+	if err != nil {
+		return err
+	}
+
+	if networkPriority != 0 {
+		err = d.state.Firewall.InstanceSetupNetPrio(d.inst.Project().Name, d.inst.Name(), veth, networkPriority)
+		if err != nil {
+			return fmt.Errorf("Failed to setup instance device network priority: %s", err)
```
Comment (tomponline) [inline lxd/device/device_utils_network.go:579]: %w here rather than %s

File: lxd/device/device_utils_network.go [hunk 13]
```diff
@@ -486,7 +486,7 @@ func networkNICRouteDelete(routeDev string, routes ...string) {
 }
 
 // networkSetupHostVethLimits applies any network rate limits to the veth device specified in the config.
-func networkSetupHostVethLimits(m deviceConfig.Device) error {
+func networkSetupHostVethLimits(d *deviceCommon, m deviceConfig.Device) error {
```
Comment (tomponline) [inline lxd/device/device_utils_network.go:489]: As we're now passing `deviceCommon` in here, we don't actually need the `m` argument and can use `d.config` instead.

Comment (tomponline) [reply lxd/device/device_utils_network.go:489]: This change should be a separate commit though to make backporting easier.

File: lxd/device/device_utils_network.go [hunk 14]
```diff
@@ -560,6 +560,40 @@ func networkSetupHostVethLimits(m deviceConfig.Device) error {
...
+		}
+	}
+
+	return nil
+}
+
+// networkClearHostVethLimits clears any network rate limits to the veth device specified in the config.
+func networkClearHostVethLimits(d *deviceCommon) error {
+	// Populate device config with volatile fields (hwaddr and host_name) if needed.
+	networkVethFillFromVolatile(d.config, d.volatileGet())
```
Comment (tomponline) [inline lxd/device/device_utils_network.go:589]: Please can you move this out of this function and put a call to it above where this function is called.

As in the future we may need to depend on that being called for multiple reasons and it would be clearer to do it in the `Stop` function for each NIC.

File: lxd/firewall/drivers/drivers_xtables.go [hunk 15]
```diff
@@ -1370,6 +1370,56 @@ func (d Xtables) InstanceClearRPFilter(projectName string, instanceName string,
 	return nil
 }
 
+// InstanceSetupNetPrio activates setting of skb->priority for the specified instance device on the host interface.
+func (d Xtables) InstanceSetupNetPrio(projectName string, instanceName string, deviceName string, netPrio int) error {
```
Comment (tomponline) [inline lxd/firewall/drivers/drivers_xtables.go:1374]: shall we make netPrio uint32 to make it clear thats the supported range (as per the validation rules later in this PR)?

File: lxd/firewall/drivers/drivers_xtables.go [hunk 16]
```diff
@@ -1370,6 +1370,56 @@ func (d Xtables) InstanceClearRPFilter(projectName string, instanceName string,
...
+		}
+	}
+
+	return nil
+}
+
+// InstanceClearNetPrio removes setting of skb->priority for the specified instance device on the host interface.
+func (d Xtables) InstanceClearNetPrio(projectName string, instanceName string, deviceName string) error {
+	if deviceName == "" {
+		return fmt.Errorf("Failed clearing netprio rules for instance (%s): device name is empty", instanceName)
```
Comment (tomponline) [inline lxd/firewall/drivers/drivers_xtables.go:1403]: I appreciate you've likely copied this from other places, but logging an instance name without a project name can be quite confusing if a LXD install has multiple projects containing the instances with the same name in multiple projects.

So really we should be doing something like this here:

```
return fmt.Errorf("Failed clearing netprio rules for instance %q in project %q: device name is empty", projectName, instanceName)
```
```

File: lxd/device/device_utils_network.go [hunk 17]
```diff
@@ -560,6 +560,40 @@ func networkSetupHostVethLimits(m deviceConfig.Device) error {
...
 
+	networkPriority := 0
+	if m["limits.priority"] != "" {
+		networkPriority, err = strconv.Atoi(m["limits.priority"])
+		if err != nil {
+			return fmt.Errorf("Failed to parse limits.priority (%s): %s", m["limits.priority"], err)
+		}
+	}
+
+	err = d.state.Firewall.InstanceClearNetPrio(d.inst.Project().Name, d.inst.Name(), veth)
```
Comment (tomponline) [inline lxd/device/device_utils_network.go:571]: I'm a little concerned that this will mean that we will make exec calls to iptables/nftables on every NIC start even if they aren't using netprio setting.

Perhaps we should update this function to accept an `oldConfig deviceConfig.Device` argument that represents the old device settings when being updated, or nil if being started, and then we can ascertain whether we need to wipe the old netprio settings before applying the new ones.

In `Update` functions this can be ascertained using `oldConfig := oldDevices[d.name]`, then we'd check for `oldConfig != nil && oldConfig["limits.priority"] != m["limits.priority"]` before clearing the netprio firewall rules.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks for this @mihalicyn I've made a few comments, and more generally:

- Please use %w when wrapping an `error` type when returning an `error`.
- We'll need an API extension commit for this, please see the style of past ones on how to do it https://github.com/canonical/lxd/commits/main/doc/api-extensions.md
- Please can you break up the per-NIC device changes into separate commit, as well as separating the structural changes (function arg changes etc) from the feature changes, to make it easier to back port.
- Testing, we should add some tests for this feature. Potentially in the test suite `container_devices_*` files for the relevant NIC, just to check that the NIC actually starts/stops OK, but we could also do with adding some nftables/xtables tests to https://github.com/canonical/lxd-ci/blob/main/tests/network-bridge-firewall script to check it works with both firewall drivers.

**Comment by tomponline** [comment]
@mihalicyn we also should update the docs and code to indicate the old setting is deprecated and I'm also thinking that if its set on an instance and the host is running cgroupv2 where it doesn't take effect we should prevent the instance starting and give an error pointing them at the new per-NIC netprio setting.

File: lxd/device/device_utils_network.go (hunk 13 above)
Comment (mihalicyn) [reply lxd/device/device_utils_network.go]: Have done. Thanks!

File: lxd/device/device_utils_network.go (hunk 14 above)
Comment (mihalicyn) [reply lxd/device/device_utils_network.go]: Have done. Thanks!

File: lxd/firewall/drivers/drivers_xtables.go (hunk 15 above)
Comment (mihalicyn) [reply lxd/firewall/drivers/drivers_xtables.go]: Done.

File: lxd/firewall/drivers/drivers_xtables.go (hunk 16 above)
Comment (mihalicyn) [reply lxd/firewall/drivers/drivers_xtables.go]: Fixed. Thanks!

File: lxd/device/device_utils_network.go (hunk 11 above)
Comment (mihalicyn) [reply lxd/device/device_utils_network.go]: Thanks!

Have fixed.

File: doc/reference/devices_nic.md (hunk 10 above)
Comment (tomponline) [reply doc/reference/devices_nic.md:445]: @mihalicyn this still needs addressing.

File: lxd/device/device_utils_network.go (hunk 17 above)
Comment (tomponline) [reply lxd/device/device_utils_network.go:571]: @mihalicyn this still needs addressing.

**Comment by tomponline** [comment]
@mihalicyn I did another pass as I couldn't remember if you had asked me to or not ;)

**Comment by tomponline** [comment]
Moving back to draft status

**Comment by mihalicyn** [comment]
> Thanks for this @mihalicyn I've made a few comments, and more generally:
> 
>     * Please use %w when wrapping an `error` type when returning an `error`.

Have done.

> 
>     * Please use %q when quoting user input or request string vars in an `error`.

+

> 
>     * We'll need an API extension commit for this, please see the style of past ones on how to do it https://github.com/canonical/lxd/commits/main/doc/api-extensions.md

Done.

> 
>     * Please can you break up the per-NIC device changes into separate commit, as well as separating the structural changes (function arg changes etc) from the feature changes, to make it easier to back port.

Done.

> 
>     * Testing, we should add some tests for this feature. Potentially in the test suite `container_devices_*` files for the relevant NIC, just to check that the NIC actually starts/stops OK, but we could also do with adding some nftables/xtables tests to https://github.com/canonical/lxd-ci/blob/main/tests/network-bridge-firewall script to check it works with both firewall drivers.

Have added additional checks to nic_bridged test.

**Comment by mihalicyn** [comment]
> I wonder if we should fail to start a container if the old style priority setting is specified when on a cgroupv2 system, otherwise it'll be silently not applied right?

I'm not sure that we should prevent container from starting because it can break something for our users after upgrade. I've decided to keep warning about netprio cgroup in logs but changed warning text and mentioned a new `limits.priority` feature. This feature does not look critical enough to prevent instance from starting if limit was not applied, IMHO.

File: lxd/device/device_utils_network.go (hunk 17 above)
Comment (mihalicyn) [reply lxd/device/device_utils_network.go]: Fixed. Thanks!

File: doc/reference/devices_nic.md (hunk 10 above)
Comment (mihalicyn) [reply doc/reference/devices_nic.md]: I've updated description. Thanks!

**Comment by mihalicyn** [comment]
Ok. I think that I've addressed all review comments. Now I'm waiting tests to pass and then mark this PR as ready for review.

File: doc/.sphinx/.wordlist.txt [hunk 18]
```diff
@@ -308,3 +308,5 @@ Zettabyte
 ZFS
 zpool
 zpools
+queueing
```
Comment (ru-fu) [inline doc/.sphinx/.wordlist.txt:311]:
```suggestion
```
Let's go for "queuing" instead - both are correct, but afaics aspell knows "queuing" but not "queueing".

File: doc/reference/devices_nic.md [hunk 19]
```diff
@@ -442,6 +442,7 @@ Key                     | Type    | Default           | Description
 `limits.egress`         | string  | -                 | I/O limit in bit/s for outgoing traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.ingress`        | string  | -                 | I/O limit in bit/s for incoming traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.max`            | string  | -                 | I/O limit in bit/s for both incoming and outgoing traffic (same as setting both `limits.ingress` and `limits.egress`)
+`limits.priority`       | integer | -                 | `skb->priority` (32-bit unsigned integer) for outgoing traffic. This value is used by the kernel queueing discipline (qdisc) to prioritize network packets. Effect of this value depending on a particular qdisc implementation (`SKBPRIO`, `QFQ`, etc). You need to consult with the kernel qdisc documentation before setting anything here.
```
Comment (ru-fu) [inline doc/reference/devices_nic.md:445]:
```suggestion
`limits.priority`       | integer | -                 | The `skb->priority` value (32-bit unsigned integer) for outgoing traffic, to be used by the kernel queuing discipline (qdisc) to prioritize network packets (The effect of this value depends on the particular qdisc implementation, for example, `SKBPRIO` or `QFQ`. Consult the kernel qdisc documentation before setting this value.)
```
Trying to fit this into the existing format - this will work much better when we move this to the code and have more space for explanations. :)

Comment (mihalicyn) [reply doc/reference/devices_nic.md]: Thanks a lot! Fixed!

File: doc/.sphinx/.wordlist.txt (hunk 18 above)
Comment (mihalicyn) [reply doc/.sphinx/.wordlist.txt]: Fixed, thanks!

**Review by ru-fu — COMMENTED** [review]
Looks good now - but is this only for routed NICs? 
If I read the code changes correctly, it's also for bridged and p2p, so you'll need to copy the option description to those tables as well.

File: doc/reference/devices_nic.md [hunk 20]
```diff
@@ -442,6 +442,7 @@ Key                     | Type    | Default           | Description
 `limits.egress`         | string  | -                 | I/O limit in bit/s for outgoing traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.ingress`        | string  | -                 | I/O limit in bit/s for incoming traffic (various suffixes supported, see {ref}`instances-limit-units`)
 `limits.max`            | string  | -                 | I/O limit in bit/s for both incoming and outgoing traffic (same as setting both `limits.ingress` and `limits.egress`)
+`limits.priority`       | integer | -                 | The `skb->priority` value (32-bit unsigned integer) for outgoing traffic, to be used by the kernel queuing discipline (qdisc) to prioritize network packets (The effect of this value depends on the particular qdisc implementation, for example, `SKBPRIO` or `QFQ`. Consult the kernel qdisc documentation before setting this value.)
```
Comment (tomponline) [inline doc/reference/devices_nic.md:445]: This needs to be added to each relevant NIC

File: doc/api-extensions.md [hunk 21]
```diff
@@ -2289,3 +2289,7 @@ Check the list of {doc}`configuration options </config-options>` to see which co
 ## `syslog_socket`
 
 This introduces a syslog socket that can receive syslog formatted log messages. These can be viewed in the events API and `lxc monitor`, and can be forwarded to Loki. To enable this feature, set `core.syslog_socket` to `true`.
+
+## `instances_nic_limits_priority`
+
+This introduces a new per-NIC `limits.priority` option.
```
Comment (tomponline) [inline doc/api-extensions.md:2295]: We could add here " that works with both cgroupv1 and cgroupv2 unlike the deprecated limits.network.priority instance setting which only worked with cgroupv1".

**Review by tomponline — CHANGES_REQUESTED** [review]
We need to update the `limits.network.priority` entry in the docs to state that its deprecated.

**Comment by tomponline** [comment]
@mihalicyn please can you also add tests to https://github.com/canonical/lxd-ci/blob/main/tests/network-bridge-firewall as this will actively test both firewall drivers before release.

**Comment by tomponline** [comment]
> > I wonder if we should fail to start a container if the old style priority setting is specified when on a cgroupv2 system, otherwise it'll be silently not applied right?
> 
> I'm not sure that we should prevent container from starting because it can break something for our users after upgrade. I've decided to keep warning about netprio cgroup in logs but changed warning text and mentioned a new `limits.priority` feature. This feature does not look critical enough to prevent instance from starting if limit was not applied, IMHO.

What do you think @escabo about this?

We have an instance level setting called `limits.network.priority` which only works with cgroupv1.
Currently if we set this setting on a cgroupv2 system is is silently ignored and no network priority is applied.

In order to fix this we are introducing a new per-NIC (rather than per-instance) setting which works with both cgroupv1 and cgroupv2. So it provides both increased support and more flexibility than the old setting.

We are planning to mark the old instance level setting as deprecated, and adding a warning to the `lxc warning list` output (if the user cares to look).

However we were wondering whether we should also change LXD so that if the instance level setting is incompatible with the host system (because its using cgroupv2) whether we should prevent the container from starting and show an error explaining that they should modify their instance's config to use the new NIC level settings.

@mihalicyn is correct that it may cause some instances to not start that were previously starting, but arguably if they have requested LXD apply a priority setting and LXD is unable to do so, the container will not get the network priority the user has requested and they may be unaware.

I'm not proposing we introduce this functionality to the LTS branch, only the monthly release.

File: doc/reference/devices_nic.md (hunk 20 above)
Comment (mihalicyn) [reply doc/reference/devices_nic.md:447]: Fixed. Thanks!

**Comment by mihalicyn** [comment]
> We need to update the `limits.network.priority` entry in the docs to state that its deprecated.

Have done.

**Comment by mihalicyn** [comment]
> @mihalicyn please can you also add tests to https://github.com/canonical/lxd-ci/blob/main/tests/network-bridge-firewall as this will actively test both firewall drivers before release.

Sure. I'll take a look and add test to `lxd-ci` too.

File: doc/api-extensions.md [hunk 22]
```diff
@@ -2289,3 +2289,7 @@ Check the list of {doc}`configuration options </config-options>` to see which co
 ## `syslog_socket`
 
 This introduces a syslog socket that can receive syslog formatted log messages. These can be viewed in the events API and `lxc monitor`, and can be forwarded to Loki. To enable this feature, set `core.syslog_socket` to `true`.
+
+## `instances_nic_limits_priority`
+
+This introduces a new per-NIC `limits.priority` option that works with both cgroupv1 and cgroupv2 unlike the deprecated limits.network.priority instance setting which only worked with cgroupv1.
```
Comment (ru-fu) [inline doc/api-extensions.md:2295]:
```suggestion
This introduces a new per-NIC `limits.priority` option that works with both cgroup1 and cgroup2 unlike the deprecated `limits.network.priority` instance setting, which only worked with cgroup1.
```

File: shared/instance.go [hunk 23]
```diff
@@ -278,6 +278,7 @@ var InstanceConfigKeysAny = map[string]func(value string) error{
 	// Controls how much priority to give to the instance's network requests when under load.
 	//
 	// Specify an integer between 0 and 10.
+	// This option is deprecated. You should use per-NIC `limits.priority` option instead.
```
Comment (ru-fu) [inline shared/instance.go:281]:
```suggestion
	// This option is deprecated. Use the per-NIC `limits.priority` option instead.
```

**Comment by escabo** [comment]
> What do you think @escabo about this?

I also think a warning is enough.

**Comment by tomponline** [comment]
> > What do you think @escabo about this?
> 
> I also think a warning is enough.

Thanks, will leave it as it is now.

**Comment by tomponline** [comment]
@mihalicyn if you can fix the spelling mistake urgently "cgroupv" then I can merge this.

**Comment by mihalicyn** [comment]
> @mihalicyn if you can fix the spelling mistake urgently "cgroupv" then I can merge this.

sure, but I also wanted to write extra tests with `lxd-ci` and test everything with different (3) NIC types.

**Comment by tomponline** [comment]
> > @mihalicyn if you can fix the spelling mistake urgently "cgroupv" then I can merge this.
> 
> sure, but I also wanted to write extra tests with `lxd-ci` and test everything with different (3) NIC types.

OK I'll hold off then. Thanks!

**Review by ru-fu — DISMISSED** [review]
Doc update looks good now. :)

**Comment by mihalicyn** [comment]
Ok. Now `limits.priority` works for `bridged` NICs too. This feature is only available for `nftables` firewall drivers because `xtables` lacks of having egress hooks for device or bridge-level hook that allows us filtering based on the input device name. After discussion with Tom we've decided to add special check to the `limits.priority` handler that prevents this option from being silently ignored if something is not supported.

Taking this change into account now I'm not sure that I need to change anything in the https://github.com/canonical/lxd-ci/blob/main/tests/network-bridge-firewall test because this test is about checking that everything works equally for nftables/xtables. For `limits.priority` it's not the case, unfortunately.

**Comment by mihalicyn** [comment]
I've rebased PR and resolved conflicts.

**Review by tomponline — APPROVED** [review]
LGTM!

**Comment by tomponline** [comment]
@mihalicyn we should just add a check for nftables in lxd-ci bridge firewall now.

**Comment by mihalicyn** [comment]
> @mihalicyn we should just add a check for nftables in lxd-ci bridge firewall now.

sure, I think we discuss approach to this thing during our regular call ;-)
